q2cli/core/usage.py

```diff
@@ -22,13 +22,25 @@ def __init__(self):
         super().__init__()
         self._cache_recorder = []
 
+    @property
     def cache(self):
         return self._cache_recorder
 
+    @classmethod
+    def from_cache(cls, cache):
+        instance = cls()
+        instance._cache_recorder = cache
+        return instance
+
     def _add_cache_record(self, source, value):
         record = dict(source=source, value=value)
         self._cache_recorder.append(record)
 
+    def render(self):
+        renderer = CLIRenderer(self._cache_recorder)
+        for rendered in renderer.render():
+            yield rendered
+
```

Comment:
What do you think about returning a string here, as `ArtifactAPIUsage.render` does?

```suggestion
    def render(self):
        renderer = CLIRenderer(self._cache_recorder)
        rendered = renderer.render()
        return '\n'.join(list(rendered))


```

Reply:
I think we should instead think about inverting the question:

What do you think about updating `ArtifactAPIUsage.render` to return a generator, like we're doing here?

Reply:
The reason I ask this is because `ArtifactAPIUsage` was the first real driver we implemented, so we didn't have as much experience as we did when we got around to implementing the CLIUsage driver.

Reply:
Sounds good to me, but it's not obvious to me why we would prefer to return a generator over a string. Maybe that will become clear to me soon.

Reply:
I think its certainly a bit subjective, but, generators are usually a good pick for handling data structures that might require lots of iteration (like these usage examples). This allows the caller to control the flow rate of the data, vs trying to drink from a firehose. If you only need the first few elements, then you can get that using a generator, without incurring the penalty of processing everything.

https://wiki.python.org/moin/Generators
https://stackoverflow.com/a/102632

Reply:
Also, a string might not be the right data type for every interface. Think q2studio - a potential driver there might render PNG images of the steps.